### PR TITLE
Send APIs

### PR DESCRIPTION
**Pull request by kspearrin**
- Alter database schema, sprocs, repositories, and supporting models for new Send table
- Add Send CRUD and public access APIs
- Services for Azure storage implementation. Still need to add local storage implementation
- Job for cleaning up deleted sends

### CODE REVIEW

File: src/Core/Models/Api/SendFileModel.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+namespace Bit.Core.Models.Api
+{
+    public class SendFileModel
+    {
+        public SendFileModel() { }
+
+        public SendFileModel(SendFileData data, GlobalSettings globalSettings)
+        {
+            Id = data.Id;
+            Url = $"{globalSettings.Send.BaseUrl}/{data.Id}";
```
Comment (cscharf) [inline src/Core/Models/Api/SendFileModel.cs:13]: Ah, I see where the URL is used now... although I still don't see a path that would match what's in the configuration JSON.

File: src/Core/Models/Table/Send.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+namespace Bit.Core.Models.Table
+{
+    public class Send : ITableObject<Guid>
+    {
+        public Guid Id { get; set; }
+        public Guid? UserId { get; set; }
+        public Guid? OrganizationId { get; set; }
+        public SendType Type { get; set; }
+        public string Data { get; set; }
+        public string Key { get; set; }
```
Comment (cscharf) [inline src/Core/Models/Table/Send.cs:14]: Correct me if I'm wrong, but isn't the key only kept client-side, or is this a different key?

File: src/Core/Services/Implementations/LocalSendStorageService.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            {
+                File.Delete(path);
+            }
+        }
+
+        private Task InitAsync()
+        {
+            if (!Directory.Exists(_baseDirPath))
+            {
+                Directory.CreateDirectory(_baseDirPath);
```
Comment (cscharf) [inline src/Core/Services/Implementations/LocalSendStorageService.cs:56]: Are there any ownership/access permissions that need to be set on the directory when creating it, `chown`, etc.? I know we've had to manage this in the past with other directories if memory serves.

File: src/Sql/Sql.sqlproj [hunk 4]
```diff
@@ -274,5 +274,18 @@
...
     <Build Include="dbo\Stored Procedures\SsoUser_ReadById.sql" />
+    <Build Include="dbo\Stored Procedures\Cipher_DeleteByIdsOrganizationId.sql" />
+    <Build Include="dbo\Stored Procedures\Cipher_SoftDeleteByIdsOrganizationId.sql" />
+    <Build Include="dbo\Stored Procedures\Organization_ReadByIdentifier.sql" />
+    <Build Include="dbo\Stored Procedures\Send_Create.sql" />
+    <Build Include="dbo\Stored Procedures\Send_DeleteById.sql" />
+    <Build Include="dbo\Stored Procedures\Send_ReadById.sql" />
+    <Build Include="dbo\Stored Procedures\Send_ReadByUserId.sql" />
+    <Build Include="dbo\Stored Procedures\Send_Update.sql" />
+    <Build Include="dbo\Stored Procedures\SsoConfig_ReadManyByNotBeforeRevisionDate.sql" />
```
Comment (cscharf) [inline src/Sql/Sql.sqlproj:285]: `SsoConfig_...`? Seems off. Was this missing from the project originally (which would have been my fault of course)?

File: src/Sql/dbo/Stored Procedures/Send_Update.sql [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    SET
+        [UserId] = @UserId,
+        [OrganizationId] = @OrganizationId,
+        [Type] = @Type,
+        [Data] = @Data,
+        [Key] = @Key,
+        [Password] = @Password,
+        [MaxAccessCount] = @MaxAccessCount,
+        [AccessCount] = @AccessCount,
+        [CreationDate] = @CreationDate,
```
Comment (cscharf) [inline src/Sql/dbo/Stored Procedures/Send_Update.sql:31]: Would you ever update `[CreationDate]`, seems like an issue, I didn't look at the other update sprocs to see how they behave but seems we should leave this off of the `UPDATE` and reserve only for `INSERT`

File: src/Sql/dbo/Stored Procedures/User_UpdateStorage.sql [hunk 6]
```diff
@@ -34,16 +35,30 @@ BEGIN
...
+    SELECT
+        @SendStorage = SUM([CTE].[Size])
+    FROM
+        [CTE]
+
     UPDATE
         [dbo].[User]
     SET
-        [Storage] = @Storage,
+        [Storage] = (@AttachmentStorage + @SendStorage),
```
Comment (cscharf) [inline src/Sql/dbo/Stored Procedures/User_UpdateStorage.sql:61]: Either of these variables could be `null` if no records exists; I would imagine causing issues when performing this addition as the result would be `null` even if one of the two measures had a valid value.

image

File: src/Sql/dbo/Stored Procedures/Organization_UpdateStorage.sql [hunk 7]
```diff
@@ -35,16 +36,31 @@ BEGIN
...
+    SELECT
+        @SendStorage = SUM([CTE].[Size])
+    FROM
+        [CTE]
+
     UPDATE
         [dbo].[Organization]
     SET
-        [Storage] = @Storage,
+        [Storage] = (@AttachmentStorage + @SendStorage),
```
Comment (cscharf) [inline src/Sql/dbo/Stored Procedures/Organization_UpdateStorage.sql:63]: Either of these variables could be `null` if no records exists; I would imagine causing issues when performing this addition as the result would be `null` even if one of the two measures had a valid value.

image

File: src/Core/Services/Implementations/SendService.cs [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                if (string.IsNullOrWhiteSpace(password))
+                {
+                    return (null, true, false);
+                }
+                var passwordResult = _passwordHasher.VerifyHashedPassword(new User(), send.Password, password);
+                if (passwordResult == PasswordVerificationResult.SuccessRehashNeeded)
+                {
+                    send.Password = HashPassword(password);
+                }
+                if (passwordResult == PasswordVerificationResult.Failed)
```
Comment (cscharf) [inline src/Core/Services/Implementations/SendService.cs:154]: `else if` vs. `if` would save a small compute cycle in this block vs. re-evaluating the value of `passwordResult` if the first condition was successful already (similar to a `switch` statement).

File: src/Core/Services/Implementations/SendService.cs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+                {
+                    return (null, false, true);
+                }
+            }
+            // TODO: maybe move this to a simple ++ sproc?
```
Comment (cscharf) [inline src/Core/Services/Implementations/SendService.cs:159]: I personally like atomic actions that do these kinds of incremental counters as close to the "source of truth" as possible (e.g. in the database), a `++` sproc might be a good option here for sure.

File: src/Sql/Sql.sqlproj (hunk 4 above)
Comment (kspearrin) [reply src/Sql/Sql.sqlproj:285]: I added a bunch of missing files to the project. I think the other guys that are not using VS keep adding them without being in the project.

File: src/Sql/dbo/Stored Procedures/Organization_UpdateStorage.sql (hunk 7 above)
Comment (kspearrin) [reply src/Sql/dbo/Stored Procedures/Organization_UpdateStorage.sql:63]: Good call, with add `ISNULL` checks

File: src/Sql/dbo/Stored Procedures/Send_Update.sql (hunk 5 above)
Comment (kspearrin) [reply src/Sql/dbo/Stored Procedures/Send_Update.sql:31]: It's on all update sprocs right now, so just following the standard. But no, you would probably never do this.

File: src/Admin/Jobs/DeleteSendsJob.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            if (!sends.Any())
+            {
+                return;
+            }
+            using (var scope = _serviceProvider.CreateScope())
+            {
+                var sendService = scope.ServiceProvider.GetRequiredService<ISendService>();
+                foreach (var send in sends)
+                {
+                    await sendService.DeleteSendAsync(send);
```
[the remark this answers is not in the document]
Comment (cscharf) [reply src/Admin/Jobs/DeleteSendsJob.cs:42]: No, let's leave it, I forgot about the file cleanup... better to keep operations grouped together per resource vs. per medium IMO, so this is good as-is.

File: src/Sql/Sql.sqlproj (hunk 4 above)
Comment (cscharf) [reply src/Sql/Sql.sqlproj:285]: I'm also guilty, will get better at that in other PR reviews as well.

**Review by cscharf — APPROVED** [review]

**Pull request merged**